zerver/actions/message_edit.py

```diff
@@ -1005,6 +1040,11 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
                     visibility_policy=new_visibility_policy,
                 )
 
+    elif (topic_name is not None or new_stream is not None) and not target_topic_has_messages:
+        sender = target_message.sender
+        if not sender.is_bot and sender not in users_losing_access:
+            apply_automatic_unmute_follow_topics_policy(sender, target_stream, target_topic_name)
```

Comment:
I think there's two things wrong with just checking `target_topic_has_messages`.
- The first is that if the stream is a private stream with protected history, we need to be checking not whether your message is now the very first message in the topic, but whether there are any visible-to-you messages in the topic. (We can't have any logic producing results visible to you depend on whether there are message in a topic on an absolute basis, or you use that to determine whether a topic exists that you don't have access to). I think that means we need to instead compute `bulk_access_stream_messages_query(sender, messages_for_topic(realm_id, new_stream.recipient_id, target_topic_name), new_stream).first()`, and compare its ID to `target_message.id`.
- This logic, as written, doesn't consider the possibility that one moves a message to be the new first message in a topic that already has messages. I hint at this a bit [here](https://github.com/zulip/zulip/pull/28980/files#r1651791539), but I discussed with @alya and she agrees that in this case, we should have the user follow the topic, to preserve the invariant that if you see a topic where user X is the first message, you can probably assume they're following that topic and you don't need to mention them, regardless of what set of moves might have gotten them there. I think the proposed logic in the previous bullet fixes this too, but wanted to flag it because we'll want the tests to verify it too.

Probably doing that access check will be more expensive/slow than would be ideal, so we probably need to also try to keep track of whether we already moved an existing followed-topic record in the `user_profiles_having_visibility_policy` logic above. This is important, because as written, if you just rename an existing topic, it's very likely that we'd end up doing a bunch of useless database queries to try to mark the first poster in that topic as following it twice within the same `do_update_messages` function call.

Reply:
>The first is that if the stream is a private stream with protected history, we need to be checking not whether your message is now the very first message in the topic, but whether there are any visible-to-you messages in the topic.

I guess this is being taken care by `users_losing_access`:
```py
# If it's moving to a private stream, all non-subscribed users are losing access
 users_losing_access = users_losing_usermessages
```

Reply:
>This logic, as written, doesn't consider the possibility that one moves a message to be the new first message in a topic that already has messages.

Makes sense, I'll update it.

Reply:
>if you just rename an existing topic, it's very likely that we'd end up doing a bunch of useless database queries to try to mark the first poster in that topic as following it twice within the same do_update_messages function call.

renaming a topic would mean `moved_all_visible_messages` will be `True`, which won't be executed under this condition as per the [CZO thread](https://chat.zulip.org/#narrow/stream/101-design/topic/moving.20topic.20to.20first.20message/near/1749049):
>We already handle renaming entire topics by moving visibility policies, so I think the goal would be for this logic to not fire in that case.